### PR TITLE
Change BlobClient.Upload to send large uploads as individual blocks

### PR DESCRIPTION
**Pull request by tg-msft**

### CODE REVIEW

File: sdk/storage/Azure.Storage.Blobs/tests/BlobClientTests.cs [hunk 1]
```diff
@@ -121,5 +121,64 @@ async Task Verify(Func<Stream, Task<Response<BlobContentInfo>>> upload)
...
+        [Category("Live")]
+        [TestCase(33 * Constants.MB)]
+        [TestCase(257 * Constants.MB)]
+        [TestCase(1 * Constants.GB)]
+        public async Task UploadAsync_LargeBlobs(long size)
+        {
+            // TODO: #6781 We don't want to add 1GB of random data in the recordings
+            if (this.Mode == RecordedTestMode.Live)
+            {
+                await this.UploadAndVerify(size, 16 * Constants.MB, Constants.MB);
```
Comment (pakrym) [inline sdk/storage/Azure.Storage.Blobs/tests/BlobClientTests.cs]: Does it work when replayed?

**Review by pakrym — APPROVED** [review]

File: sdk/storage/Azure.Storage.Blobs/src/BlobClient.cs [hunk 2]
```diff
@@ -387,14 +392,136 @@ internal BlobClient(Uri blobUri, HttpPipeline pipeline)
...
+        /// A <see cref="StorageRequestFailedException"/> will be thrown if
+        /// a failure occurs.
+        /// </remarks>
+        internal async Task<Response<BlobContentInfo>> StagedUploadAsync(
+            Stream content,
+            BlobHttpHeaders? blobHttpHeaders,
+            Metadata metadata,
+            BlobAccessConditions? blobAccessConditions,
+            IProgress<StorageProgress> progressHandler,
+            long singleBlockThreshold = BlockBlobClient.BlockBlobMaxUploadBlobBytes,
```
Comment (kfarmer-msft) [inline sdk/storage/Azure.Storage.Blobs/src/BlobClient.cs]: Do we cap singleBlockThreshold, or permit a value larger than BlockBlobMaxUploadBlobBytes?

File: sdk/storage/Azure.Storage.Common/src/StreamPartitioner.cs [hunk 3]
```diff
@@ -26,7 +26,7 @@ public StreamPartitioner(Stream stream, MemoryPool<byte> memoryPool = default)
         /// Generate a forward-only sequence of substreams based on bufferSize.
         /// </summary>
         /// <returns>StreamPartition</returns>
-        public async Task<StreamPartition> ReadAsync(int size = Constants.DefaultBufferSize, CancellationToken ct = default)
+        public async Task<StreamPartition> ReadAsync(int size = Constants.DefaultBufferSize, bool async = true, CancellationToken ct = default)
         {
             // TODO these operations should be simplified with Memory- and Span-accepting APIs in future NET Standard
```
Comment (kfarmer-msft) [inline sdk/storage/Azure.Storage.Common/src/StreamPartitioner.cs]: Is this doable now, without losing netfx 4.6.x?

**Review by kfarmer-msft — APPROVED** [review]
A couple minor questions

File: sdk/storage/Azure.Storage.Common/src/StreamPartitioner.cs (hunk 3 above)
Comment (pakrym) [reply sdk/storage/Azure.Storage.Common/src/StreamPartitioner.cs]: Yes, `System.Memory` (https://www.nuget.org/packages/System.Memory/) supports net46*

File: sdk/storage/Azure.Storage.Blobs/src/BlobClient.cs (hunk 2 above)
Comment (tg-msft) [reply sdk/storage/Azure.Storage.Blobs/src/BlobClient.cs]: It's an internal API only used for testing.  I added an assertion to make sure we never go above `BlockBlobMaxUploadBlobBytes` in `BlobClient`.

**Pull request merged**